1-Draft/RFC00XX-Module-Isolation.md

```diff
@@ -0,0 +1,251 @@
+    then `TypeResolver` should use the cache associated with ALC-M.
+
+  - Another example, if a `ITypeName` has already resolved the type via reflection in default-ALC,
+    and the `ScriptBlock` containing that `ITypeName` is now being executed in the context of an isolated module M,
+    then the cached type referred by `ITypeName` should be invalided and another resolution should be triggered.
+    Similarly, if the `ScriptBlock` has already been compiled into LINQ expression tree and have the delegate generated,
+    then the cached delegate needs to be voided and another compilation needs to be enforced to make sure type resolution is correct.
+
+For the latter category, theoretically they will continue to work without change,
+because types loaded into different load contexts are different.
+
+### Reclaim Module AssemblyLoadContext
+
+Starting from 3.0, .NET Core supports unloading an `AssemblyLoadContext` and all assemblies in it.
+Ideally, we would like to reclaim the load context associated with an isolated module when the module is removed.
+
+The cache updates called out in the [Update Type Resolution Logic](#update-type-resolution-logic) section
+were designed with the purpose of allowing the custom ALC to be reclaimed.
+By having each custom ALC to hold the cache data structure within itself,
+we won't need to hold a custom ALC or any types loaded in a custom ALC as the key or value for any caches in our engine code,
+and thus there is no need to invalidate any engine cache.
+
+However, we won't be able to guarantee that a custom ALC could be unloaded at all.
+This is because Dynamic Language Runtime (DLR) holds types in its cache, indefinitely.
+
+The DLR types live in the default-ALC,
+and it will generate a `DynamicMetaObject` object for a late-binding operation,
+which is something like a key-value pair `<restrictions, expression>` that will be held in the DLR caches.
+This is a perf improvement feature for dynamic languages like PowerShell because it allows you to skip an expensive late-binding operation when the `restrictions` matches.
+That key-value pair will be held indefinitely,
+so once types from a custom load context is involved in a late binding operation,
+such as property access, method invocation and etc,
+it will be rooted in the GC Heap by the DLR cache, and the custom load context cannot be unloaded afterwards.
+
+## Summary
+
+For the initial implementation, here are the goals and non-goals:
+
+- Goals: work items called out in the [Specification](#specification) section,
+  including updates to the module cmdlets, assembly loading, and type resolution.
+- Non-goals: the new type reference syntax.
+
+To emphasize again:
+The type identity issue is inevitable as soon as different versions of the same assembly can be loaded in the same PowerShell session.
+This will mainly affect modules that expect users to directly create objects of the types loaded by the modules.
```

Comment:
In response to @rjmholt's questions:

> How do I use `Add-Type` in a global way? Will I be able to target an ALC like with `Import-Module`?

The RFC proposes to add a new parameter `-AssemblyLoadContext` to `Import-Module`, so a user can choose to load a module using an existing module's ALC, by `Import-Module A -AssemblyLoadContext B.ReflectionContext`. (Names of the parameter and property may need to be revisited).

Can you elaborate a bit more on the `Add-Type` question? The proposal is to make `Add-Type` aware of the active ALC, so that it loads assembly into the right ALC instead of always loading into the default ALC as of today. I'm not sure what you mean by __"in a global way"__.

> How does a module author declare that their module should be loaded into its own ALC, or into the global one, or that it's designed to interoperate with another module in the same ALC? Who wins when a module author's configuration conflicts with how the user imports a module?

The discussion about this is in this thread: https://github.com/PowerShell/PowerShell-RFC/pull/298#discussion_r686427648.
My personal view is that a module author shouldn't declare where their module should be loaded, and thus there shouldn't be a configuration about this in the manifest. Please share your thoughts in that thread.

> What happens when modules are autoloaded?

The behavior proposed in this RFC would be: same as today -- by default module is not loaded in isolated context, so if there is a conflict in assembly loading, it will fail, just like today.

Loading a module in a separate ALC has some implications, including that it may result in type identity issue which would be very confusing to the user, so I don't think we should do this by default, but should always do it when user explicitly asking for it.

We could detect the error when loading a module normally, and then revert back to loading it again in a separate ALC. But again, this magic is not an explicit ask. I guess we could write out a warning message in this case, just like the `-WindowsCompat` switch does in `Import-Module`.

> Do we need new cmdlets or other mechanisms to handle/manage the new PS ALC concept? We conflate PS ALCs with modules in some places (like with the proposed typename syntax) and I feel like we should design carefully and minimally there if we can.

In this proposal, ALC is supposed to be associated with module. I don't think the goal is to allow arbitrary use of ALC.